Firestore/core/src/util/executor_std.h

```diff
@@ -62,24 +64,29 @@ class ExecutorStd : public Executor {
   Task* PopFromSchedule() override;
 
  private:
-  Id PushOnSchedule(TimePoint when, Tag tag, Operation&& operation);
+  class SharedState;
+
+  Id PushOnScheduleLocked(TimePoint when, Tag tag, Operation&& operation);
 
   void OnCompletion(Task* task) override;
   void Cancel(Id operation_id) override;
 
-  void PollingThread();
-  Id NextId();
+  static void PollingThread(std::shared_ptr<SharedState> state);
+  Id NextIdLocked();
 
-  class SharedState;
+  // A mutex that provides mutual exclusion to users of the Executor interface.
+  // Worker threads do not acquire this mutex--they only operate on the
+  // SharedState.
+  std::mutex mutex_;
+
+  std::vector<std::thread> worker_thread_pool_;
+
+  Id current_id_ = 0;
 
   // State shared with workers. Note that if the Executor's destructor is called
   // from a worker thread, this state will outlive the nominally owning
-  // Executor.
+  // Executor. `mutex_` does not protect this state.
   std::shared_ptr<SharedState> state_;
-
-  std::vector<std::thread> worker_thread_pool_;
-
-  std::atomic<Id> current_id_{0};
```

Comment:
What's the advantage of not using an atomic?

Reply:
Once we need to acquire a mutex to check whether or not the executor has been disposed, it's harder to increment not holding the lock than to just continue holding it while incrementing. At that point the atomic is a waste.